add smarts to maven binary downloader

rremer opened the pull request:
The existing tools/download-maven.sh has some problems:
1) hardcoded version (which was wrong until yesterday) - maven enforces backwards compatibility within a major revision, so whatever minor version is available is good enough
2) hardcoded tools/bin/maven/mvn path, which means if you have CI setup that already has maven of an appropriate major version, the downloader will run anyways

This updated script takes a new argument from build.sh, where the major version of maven necessary is declared as an environment variable, then download-maven.sh:
- looks on the local path to see if there's a compatible maven version
- looks upstream for a compatible version of maven
- verifies the download by sha1 (and doesn't redownload if one exists already and was uninstalled)
- unpacks and adds to local path the way the rest of mvn is invoked

I have no desire to do this in cmd or powershell (if it even exists).

Commit:
add smarts to maven binary downloader

stuartwdouglas commented:
The assumption that maven never breaks backward compatibility is incorrect, for a project this large we have had minor maven updated cause breakages in the past. 

Having different versions of maven used in the build is just asking for trouble.

rremer commented:
Um, it's written all over their documentation that they assert backwards
compatibility. Here's the maven versioning policy:
https://cwiki.apache.org/confluence/display/MAVEN/Version+number+policy
"""
If the changes since the previous tag include improvements and/or new
features which are backwards compatible with the previous tag, then the
MINOR number will be increased and the INCREMENTAL number will be reset to
0."""

Here's also where they promise to attempt backwards compatible with maven 2:
http://maven.apache.org/docs/3.1.0/release-notes.html
"""Maven 3 aims to ensure backward compatibility with Maven 2,"""

Now, in reality I've come across maven 2 projects which needed to be
restructured, but have never come across maven 3 projects which needed a
specific version of maven 3. Perhaps you're requesting a more explicit
declaration of 'this build is only asserted to succeed in minor version x
of the build tool'? We could add to the readme.

I've already worked around making this project build in Jenkins, so I'm
certainly not going to fight ya here. I have Debian packaging scripts for
you too if you want them.
On Feb 26, 2015 8:01 PM, "Stuart Douglas" <notifications@github.com> wrote:

> The assumption that maven never breaks backward compatibility is
> incorrect, for a project this large we have had minor maven updated cause
> breakages in the past.
>
> Having different versions of maven used in the build is just asking for
> trouble.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/wildfly/wildfly/pull/7204#issuecomment-76327098>.
>

jamezp commented:
They may say it's minor versions are backwards compatible and maybe with maven core that's true, it's definitely not true with all plugins though. Several plugins broke between the 3.0.x and 3.1.x versions due to the change to eclipse aether.

I don't know that our build plugins support versions less than 3.1.x due to the fact they use eclipse aether.

There's not requirement to use the `build.(sh|bat)` script so I think hard-coding is okay. Also some distro's, I can't remember which ones now, I don't think come with `sed` so we generally can't use that in our scripts.

Thank you for taking the time and thought into putting this PR together. I think we need to be consistent with the version we enforce with the build scripts. Again though using `mvn clean install` should work fine and there is no real need to use a build script. You may have to set the [JBoss maven repositories](https://developer.jboss.org/wiki/MavenGettingStarted-Users) though in your `settings.xml`.
